contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c

```diff
@@ -245,81 +289,98 @@ tdeheap_xlog_seg_write(int fd, const void *buf, size_t count, off_t offset)
+	off_t dec_off = 0;
+	size_t dec_sz = 0;
+	XLogRecPtr data_start;
+	XLogRecPtr data_end;
+
+	/* !!! TODO: should be passed as the arg? */
+	static const int wal_segsz_bytes = 16 << 20; // 16Mb
 
 #ifdef TDE_XLOG_DEBUG
-	elog(DEBUG1, "read from a WAL segment, pages amount: %d, size: %lu offset: %ld", count / (Size) XLOG_BLCKSZ, count, offset);
+	elog(DEBUG1, "read from a WAL segment, size: %lu offset: %ld [%lX], seg: %X/%X", 
+					count, offset, offset, LSN_FORMAT_ARGS(segno));
 #endif
 
+	/* 
+	 * Read data from disk
+	 */
 	readsz = pg_pread(fd, buf, count, offset);
 
-	/*
-	 * Read the buf page by page and decypt ecnrypted pages. We may start or
-	 * fihish reading from/in the middle of the page (walreceiver) in such a
-	 * case we should preserve the last read page header for the IV data and
-	 * the encryption state.
-	 *
-	 * TODO: check if walsender/receiver restarts form the beggining of the
-	 * page in case of the crash.
-	 */
-	for (dec_off = 0; dec_off < readsz;)
+	if (!keys)
 	{
-		data_size = Min(page_size, readsz);
+		/* cache is empty, try to read keys from disk */
+		keys = pg_tde_fetch_wal_keys(0);
+	}
 
-		if (page_size == XLOG_BLCKSZ)
-		{
-			memcpy((char *) curr_page_hdr, (char *) buf + dec_off, SizeOfXLogShortPHD);
+#ifndef FRONTEND
+	write_key_lsn = pg_atomic_read_u64(&EncryptionState->enc_key_lsn);
+#endif
 
-			/* set the flag to "not encrypted" for the walreceiver */
-			((XLogPageHeader) ((char *) buf + dec_off))->xlp_info &= ~XLP_ENCRYPTED;
+	if (write_key_lsn != 0)
+	{
+		WALKeyCacheRec *last_key = pg_tde_get_last_wal_key();
+		Assert(last_key);
 
-			Assert(curr_page_hdr->xlp_magic == XLOG_PAGE_MAGIC || curr_page_hdr->xlp_magic == 0);
-			dec_off += XLogPageHeaderSize(curr_page_hdr);
-			data_size -= XLogPageHeaderSize(curr_page_hdr);
-			/* it's a beginning of the page */
-			iv_ctr = 0;
-		}
-		else
+		/* write has generated a new key, need to fetch it */
+		if (last_key->start_lsn < write_key_lsn)
 		{
-			/* we're in the middle of the page */
-			iv_ctr = (offset % XLOG_BLCKSZ) - XLogPageHeaderSize(curr_page_hdr);
-		}
+			pg_tde_fetch_wal_keys(write_key_lsn);
 
-		if ((data_size + dec_off) > readsz)
-		{
-			data_size = readsz - dec_off;
+			/* in case cache was empty before */
+			keys = pg_tde_get_wal_cache_keys();
 		}
+	}
 
-		if (curr_page_hdr->xlp_info & XLP_ENCRYPTED)
+	SetXLogPageIVPrefix(tli, segno, iv_prefix);
+
+	XLogSegNoOffsetToRecPtr(segno, offset, wal_segsz_bytes, data_start);
+	XLogSegNoOffsetToRecPtr(segno, offset + count, wal_segsz_bytes, data_end);
+	
+	curr_key = keys;
+	while (curr_key)
+	{
+#ifdef TDE_XLOG_DEBUG
+		elog(DEBUG1, "WAL key %X/%X-%X/%X", LSN_FORMAT_ARGS(curr_key->start_lsn), LSN_FORMAT_ARGS(curr_key->end_lsn));
+#endif
+
+		if (curr_key->key->internal_key.start_lsn != InvalidXLogRecPtr &&
+				(curr_key->key->internal_key.rel_type & TDE_KEY_TYPE_WAL_ENCRYPTED))
 		{
-			if (key == NULL)
+			/* 
+			 * Check if the key's range overlaps with the buffer's and decypt
```

Comment:
This is the optimal case, but if we can have multiple keys in the same segment file, not necessarily true. We can turn on / off / on / off / on / off ... wal encryption, and this function will have to deal with scenarios where the specified range is encrypted by many keys in this case.

Reply:
It works with multiple keys. We decrypt everything up to the key's end_lsn or the buffer end (whatever comes first). And the go to the next key and so on and so forth.
The implementation here isn't the most efficient though. See my reply to your loop comments below.